streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java

```diff
@@ -84,34 +84,44 @@ private void initInternal(final ProcessorContext context) {
 
         cache.addDirtyEntryFlushListener(name, entries -> {
             for (final ThreadCache.DirtyEntry entry : entries) {
-                final byte[] binaryWindowKey = cacheFunction.key(entry.key()).get();
-                final long timestamp = WindowKeySchema.extractStoreTimestamp(binaryWindowKey);
-
-                final Windowed<K> windowedKey = WindowKeySchema.fromStoreKey(binaryWindowKey, windowSize, serdes.keyDeserializer(), serdes.topic());
-                final Bytes key = Bytes.wrap(WindowKeySchema.extractStoreKeyBytes(binaryWindowKey));
-                maybeForward(entry, key, windowedKey, (InternalProcessorContext) context);
-                underlying.put(key, entry.newValue(), timestamp);
+                putAndMaybeForward(entry, context);
             }
         });
     }
 
-    private void maybeForward(final ThreadCache.DirtyEntry entry,
-                              final Bytes key,
-                              final Windowed<K> windowedKey,
-                              final InternalProcessorContext context) {
+    private void putAndMaybeForward(final ThreadCache.DirtyEntry entry,
+                                    final InternalProcessorContext context) {
+        final byte[] binaryWindowKey = cacheFunction.key(entry.key()).get();
+        final Windowed<Bytes> windowedKeyBytes = WindowKeySchema.fromStoreBytesKey(binaryWindowKey, windowSize);
+        final long windowStartTimestamp = windowedKeyBytes.window().start();
+        final Bytes key = windowedKeyBytes.key();
         if (flushListener != null) {
-            final ProcessorRecordContext current = context.recordContext();
-            context.setRecordContext(entry.entry().context());
-            try {
-                final V oldValue = sendOldValues ? fetchPrevious(key, windowedKey.window().start()) : null;
-                flushListener.apply(
-                    windowedKey,
-                    serdes.valueFrom(entry.newValue()),
-                    oldValue,
-                    entry.entry().context().timestamp());
-            } finally {
-                context.setRecordContext(current);
+            final byte[] newValueBytes = entry.newValue();
+            final byte[] oldValueBytes = underlying.fetch(key, windowStartTimestamp);
+
+            // this is an optimization: if this key did not exist in underlying store and also not in the cache,
+            // we can skip flushing to downstream as well as writing to underlying store
+            if (newValueBytes != null || oldValueBytes != null) {
+                final Windowed<K> windowedKey = WindowKeySchema.fromStoreKey(windowedKeyBytes, serdes.keyDeserializer(), serdes.topic());
+                final V newValue = newValueBytes != null ? serdes.valueFrom(newValueBytes) : null;
```

Comment:
It seems that after setting `final Windowed<K> key` in the `putAndMaybeForward` method the code is more or less the same for all caching stores in question.

Just a thought, but would it be worth extracting the logic from the two similar methods and placing them in a method of `WrappedStateStore.AbstractStateStore`

Reply:
I think AbstractStateStore is not a good place to share this logic, since it is not only for wrapping caching layers, but also for other layers as well.

I feel good about keeping these three functions separated so far, mainly because their callees (fetches) are still different.